File: lib/features/settings/data/settings_repository.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/format/locale_config.dart';
import '../../../core/storage/key_value_store.dart';
import 'app_settings.dart';

abstract interface class SettingsRepository {
  AppSettings restore();
  Future<void> saveLocale(Locale? locale);
  Future<void> saveThemeMode(ThemeMode mode);
}

class StoredSettingsRepository implements SettingsRepository {
  StoredSettingsRepository(this._store);

  final KeyValueStore _store;

  static const _localeKey = 'settings.locale';
  static const _themeModeKey = 'settings.themeMode';

  @override
  AppSettings restore() {
    final language = _store.read(_localeKey);
    final themeName = _store.read(_themeModeKey);
    return AppSettings(
      localeOverride:
          language != null && localeConfigs.containsKey(language)
              ? Locale(language)
              : null,
      themeMode: ThemeMode.values.asNameMap()[themeName] ?? ThemeMode.system,
    );
  }

  @override
  Future<void> saveLocale(Locale? locale) => locale == null
      ? _store.remove(_localeKey)
      : _store.write(_localeKey, locale.languageCode);

  @override
  Future<void> saveThemeMode(ThemeMode mode) =>
      _store.write(_themeModeKey, mode.name);
}

final settingsRepositoryProvider = Provider<SettingsRepository>(
  (ref) => StoredSettingsRepository(ref.watch(keyValueStoreProvider)),
);

File: src/features/settings/data/settingsRepository.ts
import { isSupportedLanguage, type SupportedLanguage } from '@/core/format/localeConfigs';
import type { KeyValueStore } from '@/core/storage/keyValueStore';
import { type AppSettings, type ThemeMode } from './appSettings';

const localeKey = 'settings.locale';
const themeModeKey = 'settings.themeMode';
const themeModes: ThemeMode[] = ['system', 'light', 'dark'];

export interface SettingsRepository {
  restore(): AppSettings;
  saveLocale(language: SupportedLanguage | null): void;
  saveThemeMode(mode: ThemeMode): void;
}

export class StoredSettingsRepository implements SettingsRepository {
  constructor(private readonly store: KeyValueStore) {}

  restore(): AppSettings {
    const language = this.store.read(localeKey);
    const themeName = this.store.read(themeModeKey);
    return {
      localeOverride: language !== null && isSupportedLanguage(language) ? language : null,
      themeMode: themeName !== null && themeModes.includes(themeName as ThemeMode) ? (themeName as ThemeMode) : 'system',
    };
  }

  saveLocale(language: SupportedLanguage | null): void {
    if (language === null) this.store.remove(localeKey);
    else this.store.write(localeKey, language);
  }

  saveThemeMode(mode: ThemeMode): void {
    this.store.write(themeModeKey, mode);
  }
}
